### ISSUE TITLE
Add search box to market app

### ISSUE THREAD

**Issue by SergioBertolinSG** [owncloud/market#120]
Currently there is no way to search for a particular app. You have to look into the categories to find them.

With few apps this works. But when the app numbers raise, this will be a problem.

So, I suggest to add a search box similar to the one in files view to look for apps.

cc @pmaier1 @hurradieweltgehtunter

**Comment by pmaier1** [comment]
Yep, will become more important with increasing number of apps!

**Comment by hurradieweltgehtunter** [comment]
the market app will get a refactoring - backend and frontend. There is a lot we can improve. This will part of it.

**Comment by PVince81** [comment]
@pmaier1 I suggest we increase priority, there are more and more apps and it is frustrating not to have a search

**Comment by pmaier1** [comment]
> @pmaier1 I suggest we increase priority, there are more and more apps and it is frustrating not to have a search

Yes, having had a look at the other tickets this makes sense. Need to find an assignment. Moving to 'planned' as I think we want to get the next release out soon.

**Comment by PVince81** [comment]
assigned to @felixheidecke 

ownCloud already has a built-in search box which can be enabled through some APIs, I don't remember how, needs some research. Maybe have a look at the old apps page implementation which uses it.

Then wire it up with the vuejs stuff to make the search work.

**Comment by felixheidecke** [comment]
There is a thing called **STRG + F** to open the search box ;-)

**Comment by pmaier1** [comment]
> There is a thing called STRG + F to open the search box ;-)

Yes, but no :)

**Comment by PVince81** [comment]
One reason for this is that at some point there will be more than one "page" of results and the user would need to scroll down to load the next page automatically. This means that the browser's native search would not work in such cases as the next page hasn't been rendered yet.

**Comment by felixheidecke** [comment]
@pmaier1 That's not a valid answer :taco: 
@PVince81 Your's is :+1: 

We should go ahead and define how search should work. Eg. App-names only, full-text search, by license, price, include publisher or whatever. Then we can design the search around those requirements. Agreed?

**Comment by PVince81** [comment]
The more apps are added to the marketplace, the more urgent this feature is becoming.

**Comment by SamuAlfageme** [comment]
I'd also like to get the same behavior that files app has for **[ctrl/⌘ + F]** i.e. capture it & focus on the said search box. That feels really cool.

**Comment by guruz** [comment]
If there is no search input field, at least disable the lazy loading so the browser-based search function works?

**Comment by julianpoemp** [comment]
A searchbar is a standard feature of market apps. It already makes sense in a market app that offers more than 10 apps. E.g. searching for "cal" could lead to the calendar app and could offer a easy installation with one click. This is a very important feature that improves usability!

**Comment by PVince81** [comment]
Actually the marketplace API must provide a search API. Then we plug the search field, the search field connects to PHP backend in OC of market app, and PHP backend connects to marketplace to fetch and format the results.

So the prerequisite is having a search API on marketplace. Considering that the marketplace website already has a search box, the API is likely already available so we could use just that ?

**Comment by patrickjahns** [comment]
It's a "private ajax" api - unfortunately

**Comment by julianpoemp** [comment]
I checked the owncloud marketplace page. They make a easy GET call to `https://marketplace.owncloud.com/ajax/products?search=KEYWORD`. The answer is JSON. Where is the problem?

**Comment by patrickjahns** [comment]
@julianpoemp 

The ajax routes in the marketplace are used by the frontend (and considered private) - they are subject to change at any given point in time (by marketplace frontend requirements) and thus not considerable as API to be used by external services (market app). 

To have a stable system, an official API ( with versioning for changes) is required. 

Additionally:
The API from the marketplace does not consider ownCloud versions when searching. For the market app to be able to provide relevant results (as in apps allowed/working for the current ownCloud version), the search api requires the ownCloud version to be passed on. 

@PVince81 
I do not fully agree here - right now the market app fetches the full list of compatible apps via the API. The json is cached (if a cache is available) - and we could search on the result of this JSON data. 

It's a matter of perspective I guess

**Comment by julianpoemp** [comment]
@patrickjahns 
I checked the requests made from my OC installation's market place. It seems that the request to retrieve all apps is `https://PERSONAL_OWNCLOUD/index.php/apps/market/apps` and the Answer is a JSON array as well. The result should be all supported apps.
Is there no way to implement a search bar that searches in this JSON array?

**Comment by patrickjahns** [comment]
Sure there is - hence the comment

> and we could search on the result of this JSON data.

**Comment by guruz** [comment]
> If there is no search input field, at least disable the lazy loading so the browser-based search function works?

(seems done now for me, thanks)

**Comment by PVince81** [comment]
Frontend searching is only a temporary solution. Once the marketplace will contain more apps (1000+) it will not scale.

But I'd be ok for this quick workaround as it's better than no search at all.

**Comment by IljaN** [comment]
Quick Hack use this api: https://marketplace.owncloud.com/ajax/products?search=privacy, 
But I wonder if we should keep the api split and provide a additional search endpoint under /api/v1/
